File: apps/salmon-vault-android/src/com/mku/salmon/vault/main/FileAdapter.java
package com.mku.salmon.vault.main;
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.arch.core.util.Function;
import androidx.recyclerview.widget.RecyclerView;

import com.mku.func.BiConsumer;
import com.mku.salmon.vault.android.R;
import com.mku.convert.BitConverter;
import com.mku.salmon.vault.image.Thumbnails;
import com.mku.salmon.vault.utils.ByteUtils;
import com.mku.salmon.vault.utils.IPropertyNotifier;
import com.mku.salmonfs.SalmonFile;
import com.mku.utils.SalmonFileUtils;

import java.nio.charset.Charset;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingDeque;

public class FileAdapter extends RecyclerView.Adapter implements IPropertyNotifier {
    private static final String TAG = FileAdapter.class.getName();
    private static final int MAX_CACHE_SIZE = 20 * 1024 * 1024;
    private static final long VIDEO_THUMBNAIL_MSECS = 3000;
    private static final int TASK_THREADS = 4;

    private final boolean displayItems = true;
    private final List<SalmonFile> items;
    private final LayoutInflater inflater;
    private final Function<Integer, Boolean> itemClicked;
    private final Activity activity;
    private final LinkedHashMap<SalmonFile, Bitmap> bitmapCache = new LinkedHashMap<>();
    // we use a deque and add jobs to the front for better user experience
    private final LinkedBlockingDeque<ViewHolder> tasks = new LinkedBlockingDeque<>();
    private int lastPositionPressed;
    private int cacheSize = 0;
    private HashSet<SalmonFile> selectedFiles = new HashSet<>();
    private SimpleDateFormat formatter = new SimpleDateFormat("MM/dd/yyyy HH:mm:ss");
    private Mode mode = Mode.SINGLE_SELECT;
    private ExecutorService executor;
    private Runnable onCacheCleared;
    private HashSet<BiConsumer<Object, String>> observers = new HashSet<>();

    public void setOnCacheCleared(Runnable onCacheCleared) {
        this.onCacheCleared = onCacheCleared;
    }

    public FileAdapter(Activity activity, List<SalmonFile> items, Function<Integer, Boolean> itemClicked) {
        this.items = items;
        this.inflater = LayoutInflater.from(activity);
        this.itemClicked = itemClicked;
        this.activity = activity;
        createThread();
    }

    public HashSet<SalmonFile> getSelectedFiles() {
        return selectedFiles;
    }

    public void selectAll(boolean value) {
        if (value)
            selectedFiles.addAll(items);
        else
            selectedFiles.clear();
        propertyChanged(this, "SelectedFiles");
        notifyDataSetChanged();
    }

    public Mode getMode() {
        return mode;
    }

    public void setMultiSelect(boolean value) {
        setMultiSelect(value, true);
    }

    public void setMultiSelect(boolean value, boolean clear) {
        if (clear)
            selectedFiles.clear();
        mode = value ? Mode.MULTI_SELECT : Mode.SINGLE_SELECT;
        propertyChanged(this, "SelectedFiles");
        notifyDataSetChanged();
    }

    public void stop() {
        tasks.clear();
        unobservePropertyChanges();
        executor.shutdownNow();
    }

    private void createThread() {
        executor = Executors.newFixedThreadPool(TASK_THREADS);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ViewHolder viewHolder = (ViewHolder) holder;
        try {
            if (mode == Mode.MULTI_SELECT) {
                viewHolder.selected.setVisibility(View.VISIBLE);
            } else {
                viewHolder.selected.setVisibility(View.GONE);
            }
            viewHolder.salmonFile = items.get(position);
            updateSelected(viewHolder, viewHolder.salmonFile);
            updateBackgroundColor(viewHolder);
            viewHolder.filename.setText("");
            viewHolder.extension.setText("");
            viewHolder.filesize.setText("");
            viewHolder.filedate.setText("");
            if (viewHolder.salmonFile.isDirectory()) {
                viewHolder.thumbnail.setColorFilter(null);
                viewHolder.thumbnail.setImageResource(R.drawable.folder);
            } else {
                viewHolder.thumbnail.setImageBitmap(null);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        tasks.remove(viewHolder);
        tasks.addFirst(viewHolder);
        executor.submit(() -> {
            try {
                ViewHolder task = tasks.take();
                retrieveFileInfo(task);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
    }

    private void updateSelected(ViewHolder viewHolder, SalmonFile salmonFile) {
        viewHolder.selected.setChecked(selectedFiles.contains(salmonFile));
    }

    private void retrieveFileInfo(ViewHolder viewHolder) {
        long size = 0;
        long date = 0;
        String items = "";
        SalmonFile file = viewHolder.salmonFile;
        try {
            String filename = viewHolder.salmonFile.getBaseName();
            activity.runOnUiThread(() -> {
                viewHolder.filename.setText(filename);
            });
            if (viewHolder.salmonFile.isDirectory() && displayItems)
                items = viewHolder.salmonFile.getChildrenCount() + " " + activity.getString(R.string.Items);
            else
                size = viewHolder.salmonFile.getRealFile().length();
            date = viewHolder.salmonFile.getLastDateTimeModified();

            String finalItems = items;
            long finalSize = size;
            long finalDate = date;
            if (file != viewHolder.salmonFile)
                return;
            activity.runOnUiThread(() -> {
                updateFileInfo(viewHolder, filename, finalItems, viewHolder.salmonFile, finalSize, finalDate);
            });

            String ext = SalmonFileUtils.getExtensionFromFileName(filename).toLowerCase();
            if (bitmapCache.containsKey(file)) {
                activity.runOnUiThread(() -> {
                    updateIconFromCache(viewHolder, file, ext);
                });
            } else if (viewHolder.salmonFile.isFile()) {
                Bitmap bitmap = null;
                try {
                    bitmap = getFileThumbnail(viewHolder.salmonFile);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                Bitmap finalBitmap = bitmap;
                if (file != viewHolder.salmonFile)
                    return;
                activity.runOnUiThread(() -> {
                    if (finalBitmap == null) {
                        updateFileIcon(viewHolder, ext);
                    } else {
                        Animation animation = AnimationUtils.loadAnimation(activity, R.anim.thumbnail);
                        viewHolder.thumbnail.startAnimation(animation);
                        updateThumbnailIcon(viewHolder, finalBitmap);
                    }
                });
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private void updateFileInfo(ViewHolder viewHolder, String filename,
                                String items, SalmonFile salmonFile,
                                long size, long date) {
        viewHolder.filename.setText(filename);
        viewHolder.filedate.setText(formatter.format(new Date(date)));
        if (salmonFile.isDirectory()) {
            viewHolder.filesize.setText(items);
            viewHolder.extension.setText("");
            viewHolder.thumbnail.setColorFilter(null);
            viewHolder.thumbnail.setImageResource(R.drawable.folder);
        } else {
            viewHolder.thumbnail.setImageBitmap(null);
            viewHolder.extension.setText("");
            viewHolder.filesize.setText(ByteUtils.getBytes(size, 2));
        }
    }

    private boolean updateIconFromCache(ViewHolder viewHolder, SalmonFile file, String ext) {
        if (bitmapCache.containsKey(file)) {
            Bitmap bitmap = bitmapCache.get(file);
            if (bitmap == null)
                updateFileIcon(viewHolder, ext);
            else {
                updateThumbnailIcon(viewHolder, bitmapCache.get(file));
            }
            return true;
        }
        return false;
    }

    private void updateThumbnailIcon(ViewHolder viewHolder, Bitmap bitmap) {
        viewHolder.thumbnail.setImageBitmap(bitmap);
        viewHolder.thumbnail.setColorFilter(null);
        viewHolder.extension.setVisibility(View.GONE);
        viewHolder.extension.setText("");
    }

    private void updateFileIcon(ViewHolder viewHolder, String extension) {
        viewHolder.thumbnail.setImageResource(R.drawable.file);
        int extColor;
        try {
            extColor = getFileColorFromExtension(extension);
            viewHolder.thumbnail.setColorFilter(extColor, PorterDuff.Mode.MULTIPLY);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        viewHolder.extension.setVisibility(View.VISIBLE);
        viewHolder.extension.setText(extension);
    }

    private int getFileColorFromExtension(String extension) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] bytes = extension.getBytes(Charset.defaultCharset());
        byte[] hashValue = md.digest(bytes);
        StringBuilder sb = new StringBuilder();
        sb.append(BitConverter.toHex(hashValue));
        return Color.parseColor("#" + sb.substring(0, 6));
    }

    public void resetCache() {
        int reduceSize = 0;
        List<SalmonFile> keysToRemove = new LinkedList<>();
        for (SalmonFile key : bitmapCache.keySet()) {
            Bitmap bitmap = bitmapCache.get(key);
            if (bitmap != null)
                reduceSize += bitmap.getAllocationByteCount();
            if (reduceSize >= MAX_CACHE_SIZE / 2)
                break;
            keysToRemove.add(key);
        }
        for (SalmonFile key : keysToRemove) {
            Bitmap bitmap = bitmapCache.remove(key);
            if (bitmap != null)
                cacheSize -= bitmap.getAllocationByteCount();
        }
        onCacheCleared.run();
    }

    private Bitmap getFileThumbnail(SalmonFile salmonFile) throws Exception {
        Bitmap bitmap = null;
        String ext = SalmonFileUtils.getExtensionFromFileName(salmonFile.getBaseName()).toLowerCase();
        if (ext.equals("mp4")) {
            bitmap = Thumbnails.getVideoThumbnail(salmonFile, VIDEO_THUMBNAIL_MSECS);
        } else if (ext.equals("png") || ext.equals("jpg") || ext.equals("bmp") || ext.equals("webp") || ext.equals("gif")) {
            bitmap = Thumbnails.getImageThumbnail(salmonFile);
        }
        checkCacheSize();
        addBitmapToCache(salmonFile, bitmap);
        return bitmap;
    }

    private void addBitmapToCache(SalmonFile file, Bitmap bitmap) {
        bitmapCache.put(file, bitmap);
        if (bitmap != null)
            cacheSize += bitmap.getAllocationByteCount();
    }

    private void checkCacheSize() {
        if (cacheSize > MAX_CACHE_SIZE) {
            resetCache();
        }
    }

    @NonNull
    public FileAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.file_item, parent, false);
        return new ViewHolder(view, itemClicked, this);
    }

    public int getPosition() {
        return lastPositionPressed;
    }

    private void updateBackgroundColor(ViewHolder viewHolder) {
        if (viewHolder.selected.isChecked())
            viewHolder.itemView.setBackgroundColor(activity.getColor(R.color.colorPrimaryDark));
        else
            viewHolder.itemView.setBackgroundColor(0);
    }

    @Override
    public HashSet<BiConsumer<Object, String>> getObservers() {
        return observers;
    }

    public enum Mode {
        SINGLE_SELECT, MULTI_SELECT
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public ImageView thumbnail;
        public TextView filename;
        public TextView filesize;
        public TextView filedate;
        public TextView extension;
        public CheckBox selected;
        public SalmonFile salmonFile;

        public ViewHolder(View itemView, Function<Integer, Boolean> itemClicked, FileAdapter adapter) {
            super(itemView);
            thumbnail = itemView.findViewById(R.id.thumbnail);
            filename = itemView.findViewById(R.id.filename);
            filesize = itemView.findViewById(R.id.filesize);
            filedate = itemView.findViewById(R.id.filedate);
            extension = itemView.findViewById(R.id.extension);
            selected = itemView.findViewById(R.id.selected);

            itemView.setOnClickListener((View view) ->
            {
                if (mode == Mode.MULTI_SELECT) {
                    selected.setChecked(!selected.isChecked());
                    SalmonFile salmonFile = items.get(super.getLayoutPosition());
                    if (selected.isChecked())
                        selectedFiles.add(salmonFile);
                    else selectedFiles.remove(salmonFile);
                    adapter.propertyChanged(this, "SelectedFiles");
                    updateBackgroundColor(this);
                } else if (itemClicked == null || !itemClicked.apply(super.getLayoutPosition())) {
                    adapter.lastPositionPressed = super.getLayoutPosition();
                    itemView.showContextMenu();
                }
            });

            itemView.setOnLongClickListener((View view) -> {
                adapter.lastPositionPressed = super.getLayoutPosition();
                itemView.showContextMenu();
                return true;
            });
        }
    }
}

File: apps/salmon-vault-dotnet-android/src/Main/FileAdapter.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
using Android.Graphics;
using Android.Views;
using AndroidX.RecyclerView.Widget;
using Salmon.Vault.Utils;
using Salmon.Vault.Image;
using Java.Util.Concurrent;
using System.Collections.Concurrent;
using Android.Views.Animations;
using Java.Text;
using Java.Security;
using Mku.Utils;
using Mku.SalmonFS;
using BitConverter = Mku.Convert.BitConverter;
using Salmon.Vault.DotNetAndroid;
using System;
using Android.App;
using System.Collections.Generic;
using Salmon.Vault.Extensions;
using Runnable = Java.Lang.Runnable;
using Android.Widget;
using System.ComponentModel;

namespace Salmon.Vault.Main;

public class FileAdapter : RecyclerView.Adapter, INotifyPropertyChanged
{
    private static readonly string TAG = typeof(FileAdapter).Name;
    private const int MAX_CACHE_SIZE = 20 * 1024 * 1024;
    private const long VIDEO_THUMBNAIL_MSECS = 3000;
    private static readonly int TASK_THREADS = 4;

    private bool displayItems = true;
    private List<SalmonFile> items;
    private LayoutInflater inflater;
    private Func<int, bool> itemClicked;
    private Activity activity;
    private ConcurrentDictionary<SalmonFile, Bitmap> bitmapCache = new ConcurrentDictionary<SalmonFile, Bitmap>();
    // we use a deque and add jobs to the front for better user experience
    private LinkedBlockingDeque tasks = new LinkedBlockingDeque();
    private int lastPositionPressed;
    private int cacheSize = 0;
    public HashSet<SalmonFile> SelectedFiles { get; } = new HashSet<SalmonFile>();
    private SimpleDateFormat formatter = new SimpleDateFormat("MM/dd/YYYY");
    private Mode mode = Mode.SINGLE_SELECT;
    private IExecutorService executor;
    public event EventHandler OnCacheCleared;
    public event PropertyChangedEventHandler PropertyChanged;

    public FileAdapter(Activity activity, List<SalmonFile> items, Func<int, bool> itemClicked)
    {
        this.items = items;
        this.inflater = LayoutInflater.From(activity);
        this.itemClicked = itemClicked;
        this.activity = activity;
        CreateThread();
    }

    public HashSet<SalmonFile> GetSelectedFiles()
    {
        return SelectedFiles;
    }

    public void SelectAll(bool value)
    {
        if (value)
        {
            foreach (SalmonFile item in items)
            {
                SelectedFiles.Add(item);
            }
        }
        else
            SelectedFiles.Clear();
        PropertyChanged(this, new PropertyChangedEventArgs("SelectedFiles"));
        NotifyDataSetChanged();
    }

    public Mode GetMode()
    {
        return mode;
    }

    public void SetMultiSelect(bool value, bool clear = true)
    {
        if(clear)
            SelectedFiles.Clear();
        mode = value ? Mode.MULTI_SELECT : Mode.SINGLE_SELECT;
        PropertyChanged(this, new PropertyChangedEventArgs("SelectedFiles"));
        NotifyDataSetChanged();
    }

    public void Stop()
    {
        tasks.Clear();
        executor.ShutdownNow();
    }

    private void CreateThread()
    {
        executor = Executors.NewFixedThreadPool(TASK_THREADS);
    }

    public override int ItemCount => items.Count;

    public override void OnBindViewHolder(RecyclerView.ViewHolder holder, int position)
    {
        ViewHolder viewHolder = (ViewHolder)holder;
        try
        {
            if (mode == Mode.MULTI_SELECT)
            {
                viewHolder.selected.Visibility = ViewStates.Visible;
            }
            else
            {
                viewHolder.selected.Visibility = ViewStates.Gone;
            }
            viewHolder.salmonFile = items[position];
            UpdateSelected(viewHolder, viewHolder.salmonFile);
            UpdateBackgroundColor(viewHolder);
            viewHolder.filename.Text = "";
            viewHolder.extension.Text = "";
            viewHolder.filesize.Text = "";
            viewHolder.filedate.Text = "";
            if (viewHolder.salmonFile.IsDirectory)
            {
                viewHolder.thumbnail.SetColorFilter(null);
                viewHolder.thumbnail.SetImageResource(Resource.Drawable.folder);
            }
            else
            {
                viewHolder.thumbnail.SetImageBitmap(null);
            }
        }
        catch (System.Exception ex)
        {
            ex.PrintStackTrace();
        }

        tasks.Remove(viewHolder);
        tasks.AddFirst(viewHolder);
        executor.Submit(new Runnable(() =>
        {
            try
            {
                ViewHolder task = (ViewHolder)tasks.Take();
                RetrieveFileInfo(task);
            }
            catch (Java.Lang.InterruptedException e)
            {
                e.PrintStackTrace();
            }
        }));
    }

    private void UpdateSelected(ViewHolder viewHolder, SalmonFile salmonFile)
    {
        viewHolder.selected.Checked = SelectedFiles.Contains(salmonFile);
    }

    private void RetrieveFileInfo(ViewHolder viewHolder)
    {
        long size = 0;
        long date = 0;
        string items = "";
        SalmonFile file = viewHolder.salmonFile;
        try
        {
            string filename = viewHolder.salmonFile.BaseName;
            activity.RunOnUiThread(() =>
            {
                viewHolder.filename.Text = filename;
            });
            if (viewHolder.salmonFile.IsDirectory && displayItems)
                items = viewHolder.salmonFile.ChildrenCount + " " + activity.GetString(Resource.String.Items);
            else
                size = viewHolder.salmonFile.RealFile.Length;
            date = viewHolder.salmonFile.LastDateTimeModified;

            string finalItems = items;
            long finalSize = size;
            long finalDate = date;
            if (file != viewHolder.salmonFile)
                return;
            activity.RunOnUiThread(() =>
            {
                UpdateFileInfo(viewHolder, filename, finalItems, viewHolder.salmonFile, finalSize, finalDate);
            });

            string ext = SalmonFileUtils.GetExtensionFromFileName(filename).ToLower();
            if (bitmapCache.ContainsKey(file))
            {
                activity.RunOnUiThread(() =>
                {
                    UpdateIconFromCache(viewHolder, file, ext);
                });
            }
            else if (viewHolder.salmonFile.IsFile)
            {
                Bitmap bitmap = null;
                try
                {
                    bitmap = GetFileThumbnail(viewHolder.salmonFile);
                }
                catch (System.Exception e)
                {
                    e.PrintStackTrace();
                }
                Bitmap finalBitmap = bitmap;
                if (file != viewHolder.salmonFile)
                    return;
                activity.RunOnUiThread(() =>
                {
                    if (finalBitmap == null)
                    {
                        UpdateFileIcon(viewHolder, ext);
                    }
                    else
                    {
                        Animation animation = AnimationUtils.LoadAnimation(activity, Resource.Animation.thumbnail);
                        viewHolder.thumbnail.StartAnimation(animation);
                        UpdateThumbnailIcon(viewHolder, finalBitmap);
                    }
                });
            }
        }
        catch (System.Exception ex)
        {
            ex.PrintStackTrace();
        }
    }

    private void UpdateFileInfo(ViewHolder viewHolder, string filename,
                                string items, SalmonFile salmonFile,
                                long size, long date)
    {
        viewHolder.filename.Text = filename;
        viewHolder.filedate.Text = formatter.Format(new Java.Util.Date(date));
        if (salmonFile.IsDirectory)
        {
            viewHolder.filesize.Text = items;
            viewHolder.extension.Text = "";
            viewHolder.thumbnail.SetColorFilter(null);
            viewHolder.thumbnail.SetImageResource(Resource.Drawable.folder);
        }
        else
        {
            viewHolder.thumbnail.SetImageBitmap(null);
            viewHolder.extension.Text = "";
            viewHolder.filesize.Text = ByteUtils.GetBytes(size, 2);
        }
    }

    private bool UpdateIconFromCache(ViewHolder viewHolder, SalmonFile file, string ext)
    {
        if (bitmapCache.ContainsKey(file))
        {
            Bitmap bitmap = bitmapCache[file];
            if (bitmap == null)
                UpdateFileIcon(viewHolder, ext);
            else
            {
                UpdateThumbnailIcon(viewHolder, bitmapCache[file]);
            }
            return true;
        }
        return false;
    }

    private void UpdateThumbnailIcon(ViewHolder viewHolder, Bitmap bitmap)
    {
        viewHolder.thumbnail.SetImageBitmap(bitmap);
        viewHolder.thumbnail.SetColorFilter(null);
        viewHolder.extension.Visibility = ViewStates.Gone;
        viewHolder.extension.Text = "";
    }

    private void UpdateFileIcon(ViewHolder viewHolder, string extension)
    {
        viewHolder.thumbnail.SetImageResource(Resource.Drawable.file);
        int extColor;
        try
        {
            extColor = GetFileColorFromExtension(extension);
            viewHolder.thumbnail.SetColorFilter(new Color(extColor), PorterDuff.Mode.Multiply);
        }
        catch (NoSuchAlgorithmException e)
        {
            e.PrintStackTrace();
        }
        viewHolder.extension.Visibility = ViewStates.Visible;
        viewHolder.extension.Text = extension;
    }

    private int GetFileColorFromExtension(string extension)
    {
        MessageDigest md = MessageDigest.GetInstance("MD5");
        byte[] bytes = System.Text.Encoding.UTF8.GetBytes(extension);
        byte[] hashValue = md.Digest(bytes);
        Java.Lang.StringBuilder sb = new Java.Lang.StringBuilder();
        sb.Append(BitConverter.ToHex(hashValue));
        return Color.ParseColor("#" + sb.Substring(0, 6));
    }

    public void ResetCache()
    {
        int reduceSize = 0;
        List<SalmonFile> keysToRemove = new List<SalmonFile>();
        foreach (SalmonFile key in bitmapCache.Keys)
        {
            Bitmap bitmap = bitmapCache[key];
            if (bitmap != null)
                reduceSize += bitmap.AllocationByteCount;
            if (reduceSize >= MAX_CACHE_SIZE / 2)
                break;
            keysToRemove.Add(key);
        }
        foreach (SalmonFile key in keysToRemove)
        {
            bitmapCache.Remove(key, out Bitmap bitmap);
            if (bitmap != null)
                cacheSize -= bitmap.AllocationByteCount;
        }
        OnCacheCleared(this,new EventArgs());
    }

    private Bitmap GetFileThumbnail(SalmonFile salmonFile)
    {
        Bitmap bitmap = null;
        string ext = SalmonFileUtils.GetExtensionFromFileName(salmonFile.BaseName).ToLower();
        if (ext.Equals("mp4"))
        {
            bitmap = Thumbnails.GetVideoThumbnail(salmonFile, VIDEO_THUMBNAIL_MSECS);
        }
        else if (ext.Equals("png") || ext.Equals("jpg") || ext.Equals("bmp") || ext.Equals("webp") || ext.Equals("gif"))
        {
            bitmap = Thumbnails.GetImageThumbnail(salmonFile);
        }
        CheckCacheSize();
        AddBitmapToCache(salmonFile, bitmap);
        return bitmap;
    }

    private void AddBitmapToCache(SalmonFile file, Bitmap bitmap)
    {
        bitmapCache[file] = bitmap;
        if (bitmap != null)
            cacheSize += bitmap.AllocationByteCount;
    }

    private void CheckCacheSize()
    {
        if (cacheSize > MAX_CACHE_SIZE)
        {
            ResetCache();
        }
    }

    public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int viewType)
    {
        View view = inflater.Inflate(Resource.Layout.file_item, parent, false);
        return new ViewHolder(view, itemClicked, this);
    }

    public int GetPosition()
    {
        return lastPositionPressed;
    }

    private void UpdateBackgroundColor(ViewHolder viewHolder)
    {
        if (viewHolder.selected.Checked)
            viewHolder.ItemView.SetBackgroundColor(new Color(activity.GetColor(Resource.Color.colorPrimary)));
        else
            viewHolder.ItemView.SetBackgroundColor(new Color(0));
    }

    public enum Mode
    {
        SINGLE_SELECT, MULTI_SELECT
    }

    public class ViewHolder : RecyclerView.ViewHolder
    {
        private FileAdapter adapter;
        public ImageView thumbnail;
        public TextView filename;
        public TextView filesize;
        public TextView filedate;
        public TextView extension;
        public CheckBox selected;
        public SalmonFile salmonFile;

        public ViewHolder(View itemView, Func<int, bool> itemClicked, FileAdapter adapter) : base(itemView)
        {
            this.adapter = adapter;
            thumbnail = (ImageView)itemView.FindViewById(Resource.Id.thumbnail);
            filename = (TextView)itemView.FindViewById(Resource.Id.filename);
            filesize = (TextView)itemView.FindViewById(Resource.Id.filesize);
            filedate = (TextView)itemView.FindViewById(Resource.Id.filedate);
            extension = (TextView)itemView.FindViewById(Resource.Id.extension);
            selected = (CheckBox)itemView.FindViewById(Resource.Id.selected);

            itemView.Click += (object sender, EventArgs e) =>
            {
                if (adapter.mode == Mode.MULTI_SELECT)
                {
                    selected.Checked = !selected.Checked;
                    SalmonFile salmonFile = adapter.items[base.LayoutPosition];
                    if (selected.Checked)
                        adapter.SelectedFiles.Add(salmonFile);
                    else adapter.SelectedFiles.Remove(salmonFile);
                    adapter.PropertyChanged(this, new PropertyChangedEventArgs("SelectedFiles"));
                    adapter.UpdateBackgroundColor(this);
                }
                else if (itemClicked == null || !itemClicked(base.LayoutPosition))
                {
                    adapter.lastPositionPressed = base.LayoutPosition;
                    itemView.ShowContextMenu();
                }
            };

            itemView.LongClick += (object sender, View.LongClickEventArgs e) =>
            {
                adapter.lastPositionPressed = base.LayoutPosition;
                itemView.ShowContextMenu();
            };
        }
    }
}